### PR TITLE
Apply dsv3 router gemm kernel for deepseek-r1 fp4

### PR DESCRIPTION
**Pull request by Fridge003**
## Motivation

Apply kernel added in #7627 to dpsk-r1 fp4 model.

Currently we have to convert the dtype of router logits back to bf16, otherwise there will be cuda illegal memory bug. We should modify the router gemm kernel so that it outputs bf16 logits.

This PR can improve performance on bs = 1, but for larger bs like 4 or 16, the performance will drop due to the extra dtype conversions.

## Accuracy Test
```bash
python3 benchmark/gsm8k/bench_sglang.py --num-shots 8 --num-questions 1319 --parallel 1319

Accuracy: 0.904
```

## Profile Result
Profile with bs=1:
```bash
python3 -m sglang.bench_serving --model /dev/shm/DeepSeek-R1-FP4 --dataset-name random --backend sglang-oai --random-range-ratio 1 --random-input-len 1200 --random-output-len 20 --max-concurrency 1 --num-prompts 5 --profile
```

Before:
before_opt

After:
after_opt

## Benchmark Result

```bash
python3 -m sglang.launch_server --model-path /dev/shm/DeepSeek-R1-FP4 --trust-remote-code --quantization modelopt_fp4 --tp 8 --enable-flashinfer-moe

curl http://127.0.0.1:30000/flush_cache
python3 -m sglang.bench_serving --backend sglang-oai  --dataset-name random --random-input-len 1000 --random-output-len 1000 --random-range-ratio 0.98 --num-prompts 5 --max-concurrency 1 --output-file dsv3.jsonl --port 30000
```

Main branch:
```bash

+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|    |   max_concurrency |   input_throughput |   output_throughput |   mean_ttft_ms |   median_ttft_ms |   p99_ttft_ms |   mean_tpot_ms |   median_tpot_ms |   p99_tpot_ms |   per_user_throughput |
+====+===================+====================+=====================+================+==================+===============+================+==================+===============+=======================+
|  0 |             1.000 |            137.466 |             137.522 |        236.763 |          152.591 |       536.760 |          7.038 |            7.060 |         7.831 |               137.522 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
```

This PR:

```bash
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
|    |   max_concurrency |   input_throughput |   output_throughput |   mean_ttft_ms |   median_ttft_ms |   p99_ttft_ms |   mean_tpot_ms |   median_tpot_ms |   p99_tpot_ms |   per_user_throughput |
+====+===================+====================+=====================+================+==================+===============+================+==================+===============+=======================+
|  0 |             1.000 |            138.558 |             138.614 |        263.883 |          177.071 |       557.015 |          6.951 |            6.832 |         7.749 |               138.614 |
+----+-------------------+--------------------+---------------------+----------------+------------------+---------------+----------------+------------------+---------------+-----------------------+
```

About 1% increase in throughput.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhyncs** [comment]
ref https://github.com/sgl-project/sglang/pull/7676

**Comment by zhyncs** [comment]
@Fridge003 please rebase. btw is this pr ready for review

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -224,7 +225,18 @@ def forward(self, hidden_states):
                 True,  # is_vnni
             )
 
-        logits = F.linear(hidden_states, self.weight, None)
+        if (
+            hidden_states.shape[0] < 4
+            and hidden_states.shape[1] == 7168
+            and self.weight.shape[0] == 256
+            and get_device_sm() >= 90
```
Comment (zhyncs) [inline python/sglang/srt/models/deepseek_v2.py]: we don't need to invoke this every time

**Review by zhyncs — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (Fridge003) [reply python/sglang/srt/models/deepseek_v2.py]: fixed

**Review by zhyncs — APPROVED** [review]
